File: test/system/auto/stress/table.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

import os
from TestUtils import TestUtilsMixin, ROOT, ROOT_PASSWORD, ACCUMULO_DIR

class Table(TestUtilsMixin, unittest.TestCase):
    "create and destroy the same table repeatedly, sometimes ingesting into it"

    order = 85

    def deleteTable(self, table):
        out, err, code = self.rootShell(self.masterHost(),
                                        "deletetable %s\n" % table)
        self.processResult(out, err, code)
        self.sleep(0.5)
        
    def tables(self):
        code, out, err = self.shell(self.masterHost(), "tables\n")
        self.processResult(out, err, code)
        return out

    def runTest(self):
        import tempfile
        fileno, self.filename = tempfile.mkstemp()
        fp = os.fdopen(fileno, "wb")
        try:
            for i in range(0, 200, 10):
                fp.write("%08x\n" % (i * 1000))
            fp.close()

            for i in range(5):
                self.createTable('test_ingest')
                self.deleteTable('test_ingest')
            self.createTable('test_ingest')
            for i in range(5):
                self.wait(self.ingest(self.masterHost(), 10, start=i*10))
                self.wait(self.verify(self.masterHost(), 10, start=i*10))
                self.deleteTable('test_ingest')
                self.createTable('test_ingest', self.filename)
        finally:
            os.unlink(self.filename)
        
def suite():
    result = unittest.TestSuite()
    result.addTest(Table())
    return result

File: core/src/main/java/org/apache/accumulo/core/client/impl/Table.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.accumulo.core.client.impl;

import java.util.concurrent.ExecutionException;

import org.apache.accumulo.core.client.Instance;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;

public class Table {

  /**
   * Object representing an internal table ID. This class was created to help with type safety. For help obtaining the value of a table ID from Zookeeper, see
   * {@link Tables#getTableId(Instance, String)}
   *
   * Uses an internal cache and private constructor for storing a WeakReference of every Table.ID. Therefore, a Table.ID can't be instantiated outside this
   * class and is accessed by calling Table.ID.{@link #of(String)}.
   */
  public static class ID extends AbstractId {
    private static final long serialVersionUID = 7399913185860577809L;
    static final Cache<String,ID> cache = CacheBuilder.newBuilder().weakValues().build();

    public static final ID METADATA = of("!0");
    public static final ID REPLICATION = of("+rep");
    public static final ID ROOT = of("+r");

    private ID(final String canonical) {
      super(canonical);
    }

    /**
     * Get a Table.ID object for the provided canonical string.
     *
     * @param canonical
     *          table ID string
     * @return Table.ID object
     */
    public static ID of(final String canonical) {
      try {
        return cache.get(canonical, () -> new Table.ID(canonical));
      } catch (ExecutionException e) {
        throw new AssertionError("This should never happen: ID constructor should never return null.");
      }
    }
  }

}
